File: Lab 1 - C parser/parser.c
/**********************************************************************/
/* lab 1 DVG C01 - Parser OBJECT                                      */
/**********************************************************************/

/**********************************************************************/
/* Include files                                                      */
/**********************************************************************/
#include <stdio.h>
#include <ctype.h>
#include <string.h>
// #include "parser.h"



/**********************************************************************/
/* Other OBJECT's METHODS (IMPORTED)                                  */
/**********************************************************************/
#include "keytoktab.h"          /* when the keytoktab is added   */
#include "lexer.h"              /* when the lexer     is added   */
#include "symtab.h"             /* when the symtab    is added   */
#include "optab.h"             /* when the optab     is added   */

/**********************************************************************/
/* OBJECT ATTRIBUTES FOR THIS OBJECT (C MODULE)                       */
/**********************************************************************/
#define DEBUG 0
static int  lookahead=0;
static int  is_parse_ok=1;

#define char👉 char*
/**********************************************************************/
/* RAPID PROTOTYPING - simulate the token stream & lexer (get_token)  */
/**********************************************************************/
/* define tokens + keywords NB: remove this when keytoktab.h is added */
/**********************************************************************/
//enum tvalues { program = 257, id, input, output, var, begin, end, number, integer, boolean, real, coleq};
/**********************************************************************/
/* Simulate the token stream for a given program                      */
/**********************************************************************/
// static int tokens[] = { program, id, '(', input, ',', output, ')', ';', 
//     var, id, ',', id, ',', id, ':', integer, ';', begin, 
//     id, assign, id, '+', id, '*', number, end, '.', 
//     '$'};

/**********************************************************************/
/*  Simulate the lexer -- get the next token from the buffer          */
/**********************************************************************/
// static int pget_token()
// {
//     static int i=0;
//     if (tokens[i] != '$') return tokens[i++]; else return '$';
// }

/**********************************************************************/
/*  PRIVATE METHODS for this OBJECT  (using "static" in C)            */
/**********************************************************************/
static void in(char* s)
{
    if(DEBUG) printf("\n *** In  %s", s);
}
static void out(char* s)
{
    if(DEBUG) printf("\n *** Out %s", s);
}
/**********************************************************************/
/* The Parser functions                                               */
/**********************************************************************/
static void match(int t)
{
    if(DEBUG) printf("\n ------ In  match		expected %s found %s",
                    tok2lex(t), tok2lex(lookahead));
    if (lookahead == t) lookahead = get_token();
    else {
    is_parse_ok=0;
    // printf("\n *** Unexpected Token: expected: %s found: %s (in match)",
    //           tok2lex(t), tok2lex(lookahead));
    if (t == id) {
        printf("\nSYNTAX:\tSymbol expected id found %s", get_lexeme());
    } else {
        printf("\nSYNTAX:\tSymbol expected %s found %s", tok2lex(t), get_lexeme());
        }
    }
}

/**********************************************************************/
/* The grammar functions                                              */
/**********************************************************************/

static void var_part();
static void var_dec_list();
static void var_dec();
static void id_list();
static void type();
static void stat_part();
static void stat_list();
static void stat();
static void assign_stat();
static toktyp expr();
static toktyp term();
static toktyp factor();
static toktyp operand();

static void program_header()
{
    in("program_header");
    match(program);
    if (lookahead == id) {
        addp_name(get_lexeme());
    } else {
        addp_name("???");
        // printf("\nSYNTAX:   ID expected found %s", get_lexeme());
        is_parse_ok = 0;
    }
    match(id);
     match('('); match(input); match(','); match(output); match(')'); match(';');

    out("program_header");
}

static void var_part()
{
    in("var_part");
    match(var); var_dec_list();
    out("var_part");
}

static void var_dec_list()
{
    in("var_dec_list");
    var_dec();
    if (lookahead == id)
        {
            var_dec_list();
        }
    out("var_dec_list");
}

static void var_dec()
{
    in("var_dec");
    id_list(); match(':'); type(); match(';');
    out("var_dec");
}

static void id_list()
{
    in("id_list");
    char👉 lexme = get_lexeme();
    if (lookahead == id) {
        if (find_name(lexme)){
            is_parse_ok = 0; // Are we allowed to touch this?
            printf("\nSEMANTIC: ID already declared: %s", lexme);
        } else {
            addv_name(get_lexeme());
        }
        match(id);

    } else {
        is_parse_ok = 0;
        printf("\nSYNTAX:\tSymbol expected id found %s", lexme);
    }
    
    if (lookahead == ',')
    {
        match(','); id_list();
    }
    out("id_list");
}

static void type()
{
    in("type");
    switch(lookahead)
    {
        case integer:
            match(integer);
            setv_type(integer);
            break;
        case real:
            match(real);
            setv_type(real);
            break;
        case boolean:
            match(boolean);
            setv_type(boolean);
            break;
        default:
            is_parse_ok = 0;
            printf("\nSYNTAX:\tType name expected found  %s", get_lexeme());
            setv_type(error);
            break;
    }
    out("type");
}

static void stat_part()
{
    in("static_part");
    match(begin); stat_list(); match(end); match('.');

    
    out("static_part");
}

static void stat_list()
{
    in("stat_list");
    stat();
    if (lookahead == ';')
    {
        match(';'); stat_list();
    }
    out("stat_list");
}

static void stat()
{
    in("stat");
    assign_stat();
    out("stat");
}

static void assign_stat()
{
    in("assign_stat");
    toktyp left_type;
    toktyp right_type;
    
    char👉 lexme = get_lexeme();
    left_type = get_ntype(lexme);
    if (!find_name(lexme) && lookahead == id) {
        printf("\nSEMANTIC: ID NOT declared: %s", lexme);
        is_parse_ok = 0;
    }
    match(id); 
    match(assign);
    if (lookahead == '+' || lookahead == '*') {
        is_parse_ok = 0;
        printf("\nSYNTAX:\tOperand expected");
    }
    right_type = expr(); 

    if (right_type == nfound) {
        is_parse_ok = 0;
        printf("\nSYNTAX:\tOperand expected");
    }
    if (left_type != right_type) {
        // is_parse_ok = 0;

        // NOTE: i want this to throw error here but we get diff to expected output if we do that.
        printf("\nSEMANTIC: Assign types: %s := %s", tok2lex(left_type), tok2lex(right_type));
    }

    
    out("assign_stat");
}

static toktyp expr()
{
    in("expr");
    toktyp left = term();
    if (lookahead == '+')
        {
            match('+');
            toktyp right = expr();
            return get_otype('+', left, right);
        }
    return left;
    out("expr");
}

static toktyp term()
{
    in("term");
    // factor();
    toktyp left = factor();
    if (lookahead == '*')
        {
            match('*'); 
            toktyp right = term();
            return get_otype('*', left, right);
        }
    return left;
    out("term");
}

static toktyp factor()
{
    in("factor");
    toktyp type;
    if (lookahead == '(')
        {
            match('('); type = expr(); match(')');
        }
    else
        type = operand();
        
    return type;
    out("factor");
}

static toktyp operand()
{
    in("operand");
    toktyp type = nfound;
    if (lookahead == id)
    {
        char👉 lexme = get_lexeme();
        if (!find_name(lexme)) {
            is_parse_ok = 0;
            printf("\nSEMANTIC: ID NOT declared: %s", lexme);
            type = nfound;
        } else {
            type = get_ntype(lexme);
        }
        match(id);
    }
    else if(lookahead == number)
    {
        type = integer;
        match(number);
    }
    return type;
    out("operand");
}



/**********************************************************************/
/*  PUBLIC METHODS for this OBJECT  (EXPORTED)                        */
/**********************************************************************/

int parser()
{
    in("parser");


    lookahead = get_token();       // get the first token

    // while (isspace(lookahead)) get_token();

    if (lookahead == '$') {
        is_parse_ok = 0;
        printf("\nWARNING:  Input file is empty");
    } else {
        program_header();               // call the first grammar rule
        var_part();
        stat_part();
    }

    if (lookahead != '$') {
        printf("\nSYNTAX:\tExtra symbols after end of parse!\n          ");
        is_parse_ok = 0;
        printf("%s ", get_lexeme());
        while (get_token() != '$') printf("%s ", get_lexeme());
    }

    

    out("parser");
    printf("\n________________________________________________________ ");
    p_symtab();
    // if parseok do get_tokena na dpirint it?


    if (!is_parse_ok) printf("\n Parse Failed!");
    else printf("\n Parse Successful!");
    printf("\n________________________________________________________ \n");
    
    return is_parse_ok;             // status indicator
}

/**********************************************************************/
/* End of code                                                        */
/**********************************************************************/

File: Lab 4 - 🐍/parser.py
#!/usr/bin/env python3
from lexer import Lexer
from symtab import SymTab
from opttab import OpTab
from tokens import TokenType, tok2lex


class Parser:
    def __init__(self):
        self.lexer = Lexer()
        self.symtab = SymTab()
        self.optab = OpTab()
        self.lookahead = None
        self.is_parse_ok = True

    def match(self, t):
        if self.lookahead == t:
            self.lookahead = self.lexer.get_token()
        else:
            self.is_parse_ok = False
            if t == TokenType.id:
                print(f"SYNTAX:\tSymbol expected id found {self.lexer.get_lexeme()}")
            else:
                print(f"SYNTAX:\tSymbol expected {tok2lex(t)} found {self.lexer.get_lexeme()}")

    def program_header(self):
        self.match(TokenType.program)
        if self.lookahead == TokenType.id:
            self.symtab.addp_name(self.lexer.get_lexeme())
        else:
            self.symtab.addp_name("???")
            self.is_parse_ok = False
        self.match(TokenType.id)
        self.match('(')
        self.match(TokenType.input)
        self.match(',')
        self.match(TokenType.output)
        self.match(')')
        self.match(';')

    def var_part(self):
        self.match(TokenType.var)
        self.var_dec_list()

    def var_dec_list(self):
        self.var_dec()
        if self.lookahead == TokenType.id:
            self.var_dec_list()

    def var_dec(self):
        self.id_list()
        self.match(':')
        self.type_rule()
        self.match(';')

    def id_list(self):
        lexme = self.lexer.get_lexeme()
        if self.lookahead == TokenType.id:
            if self.symtab.find_name(lexme):
                self.is_parse_ok = False
                print(f"SEMANTIC: ID already declared: {lexme}")
            else:
                self.symtab.addv_name(lexme)
            self.match(TokenType.id)
        else:
            self.is_parse_ok = False
            print(f"SYNTAX:\tSymbol expected id found {lexme}")
        if self.lookahead == ',':
            self.match(',')
            self.id_list()

    def type_rule(self):
        if self.lookahead == TokenType.integer:
            self.match(TokenType.integer)
            self.symtab.setv_type(TokenType.integer)
        elif self.lookahead == TokenType.real:
            self.match(TokenType.real)
            self.symtab.setv_type(TokenType.real)
        elif self.lookahead == TokenType.boolean:
            self.match(TokenType.boolean)
            self.symtab.setv_type(TokenType.boolean)
        else:
            self.is_parse_ok = False
            print(f"SYNTAX:\tType name expected found  {self.lexer.get_lexeme()}")
            self.symtab.setv_type(TokenType.error)

    def stat_part(self):
        self.match(TokenType.begin)
        self.stat_list()
        self.match(TokenType.end)
        self.match('.')

    def stat_list(self):
        self.stat()
        if self.lookahead == ';':
            self.match(';')
            self.stat_list()

    def stat(self):
        self.assign_stat()

    def assign_stat(self):
        lexme = self.lexer.get_lexeme()
        left_type = self.symtab.get_ntype(lexme)
        if not self.symtab.find_name(lexme) and self.lookahead == TokenType.id:
            print(f"SEMANTIC: ID NOT declared: {lexme}")
            self.is_parse_ok = False
        self.match(TokenType.id)
        self.match(TokenType.assign)
        if self.lookahead in ('+', '*'):
            self.is_parse_ok = False
            print("SYNTAX:\tOperand expected")
        right_type = self.expr()
        if right_type == TokenType.nfound:
            self.is_parse_ok = False
            print("SYNTAX:\tOperand expected")
        if left_type != right_type:
            print(f"SEMANTIC: Assign types: {tok2lex(left_type)} := {tok2lex(right_type)}")

    def expr(self):
        left = self.term()
        if self.lookahead == '+':
            self.match('+')
            right = self.expr()
            return self.optab.get_otype('+', left, right)
        return left

    def term(self):
        left = self.factor()
        if self.lookahead == '*':
            self.match('*')
            right = self.term()
            return self.optab.get_otype('*', left, right)
        return left

    def factor(self):
        if self.lookahead == '(':
            self.match('(')
            t = self.expr()
            self.match(')')
            return t
        else:
            return self.operand()

    def operand(self):
        if self.lookahead == TokenType.id:
            lexme = self.lexer.get_lexeme()
            if not self.symtab.find_name(lexme):
                print(f"SEMANTIC: ID NOT declared: {lexme}")
                self.is_parse_ok = False
                t = TokenType.nfound
            else:
                t = self.symtab.get_ntype(lexme)
            self.match(TokenType.id)
            return t
        elif self.lookahead == TokenType.number:
            self.match(TokenType.number)
            return TokenType.integer
        else:
            return TokenType.nfound

    def parse(self):
        self.lookahead = self.lexer.get_token()
        if self.lookahead == '$':
            self.is_parse_ok = False
            print("WARNING:  Input file is empty")
        else:
            self.program_header()
            self.var_part()
            self.stat_part()
        if self.lookahead != '$':
            print("SYNTAX:\tExtra symbols after end of parse!")
            self.is_parse_ok = False
            print(self.lexer.get_lexeme(), end=' ')
            while self.lexer.get_token() != '$':
                print(self.lexer.get_lexeme(), end=' ')
            print()
        print("________________________________________________________ ")
        self.symtab.p_symtab()
        if not self.is_parse_ok:
            print("\n Parse Failed!")
        else:
            print("\n Parse Successful!")
        print("________________________________________________________ ")
        return self.is_parse_ok


if __name__ == "__main__":
    Parser().parse()